swagger/smart_transfer_protocol_ws.yaml

```diff
@@ -0,0 +1,332 @@
+        name: DatasetPullResponse
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - newBlocks
+          properties:
+            newBlocks:
+              $ref: '#/components/schemas/ObjectsBatch'
+  
+  dataset/pull/objects/negotiate:
+    subscribe:
+      operationId: odf.dataset.pull.objects.negotiate
+      summary: Requests to negotiate the download protocol for 1 or more object files
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPullObjectsNegotiate
+        name: DatasetPullObjectsNegotiate
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - objectFiles
+          properties:  
+            objectFiles:
+              type: array
+              description: List of object file references in this bulk
+              items:
+                $ref: '#/components/schemas/ObjectFileReference'
+      
+  dataset/pull/objects/negotiate/response:
+    publish:
+      operationId: odf.dataset.pull.objects.negotiate.response
+      summary: Response on negotiated status of object download operations during pull
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPullObjectsNegotiateResponse
+        name: DatasetPullObjectsNegotiateResponse
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - negotiatedDownloads
+          properties:   
+            negotiatedDownloads:
+              type: array
+              description: Negotiated download operations for requested bulk of object files
+              items:
+                $ref: '#/components/schemas/NegotiatedDownload'
+
+  dataset/push/request:
+    subscribe:
+      operationId: odf.dataset.push.request
+      summary: Initiates dataset push to the ODF server
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPushRequest
+        name: DatasetPushRequest
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - datasetId
+            - newBlocks
+          properties:
+            datasetId:
+              type: string
+              format: did
+              description: Dataset identifier
+            newBlocks:
+              $ref: '#/components/schemas/ObjectsBatch'
+
+  dataset/push/request/accepted:
+    publish:
+      operationId: odf.dataset.push.request.accepted
+      summary: Informs about succesfull acceptance of dataset push request
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPushRequestAccepted
+        name: DatasetPushRequestAccepted
+        contentType: application/json
+        payload:
+          type: object
+          description: Empty object indicating success
+
+  dataset/push/objects/negotiate:
```

Comment:
Word `negotiate` for me associates with some protocol phase that happens once and early on in the message exchange. Perhaps `transfer.request` and `transfer.response` are more appropriate for these steps of actually sending over data?

As I mentioned on Slack - in future we may extend the protocol to support in-band data transfer (e.g. you have a very simple ODF server and don't want to bother with pre-signed URLs), so `transfer` would be even more appropriate then.

Reply:
Agreed on this naming.
With regards to pre-signed URLs, the protocol does not mention this term, in the latest edition it looks like this:

```
    TransferUrl:
      type: object
      required:
        - url
      properties:
        url:
          type: string
          format: url
        expiresAt:
          type: string
          format: date-time
```

It could mean a pre-signed URL or just direct HTTP url. I assume `expiresAt` field would only be sent with pre-signed, but this is irrelevant for the protocol.